test/py_unit/test_lstm.py

```diff
@@ -0,0 +1,217 @@
+        output_seq_len=output_seq_len,
+        num_features=num_features,
+        stride=seq_stride,
+        time_covariates=["hour_of_day", "day_of_week"],
+        keep_last_time_cov=True,
+    )
+
+    model.to_device("cuda" if use_cuda else "cpu")
+    model.input_state_update = True
+    model.num_samples = train_dtl.dataset["value"][0].shape[0]
+    out_updater = OutputUpdater(model.device)
+
+    # -------------------------------------------------------------------------#
+    # Training
+    mses = []
+    # Initialize the sequence length
+    mu_sequence = np.ones(input_seq_len, dtype=np.float32)
+    var_y = np.full((batch_size * len(output_col),), 0.6**2, dtype=np.float32)
+
+    for _ in np.arange(2):
+        batch_iter = train_dtl.create_data_loader(batch_size, shuffle=False)
+        y_train = []
+        # for x, y in batch_iter:
+        for idx_sample, (x, y) in enumerate(batch_iter):
+
+            # replace nan in input x by the lstm_prediction:
+            if idx_sample < input_seq_len + infer_window_len:
+                x = replace_with_prediction(x, mu_sequence)
+
+            # Feed forward
+            m_pred, _ = model(x)
+
+            # Update output layer
+            out_updater.update(
+                output_states=model.output_z_buffer,
+                mu_obs=y,
+                var_obs=var_y,
+                delta_states=model.input_delta_z_buffer,
+            )
+            # Feed backward
+            model.backward()
+            model.step()
+
+            # Training metric
+            pred = normalizer.unstandardize(
+                m_pred, train_dtl.x_mean[output_col], train_dtl.x_std[output_col]
+            )
+            y_train.append(y)
+            obs = normalizer.unstandardize(
+                y, train_dtl.x_mean[output_col], train_dtl.x_std[output_col]
+            )
+            if np.isnan(obs).any():
+                continue
+            mse = metric.mse(pred, obs)
+            mses.append(mse)
+
+            # Add new prediction to mu_sequence
+            mu_sequence = np.append(mu_sequence, m_pred)
+            mu_sequence = mu_sequence[-input_seq_len:]
+
+        # Smoother
+        mu_zo_smooth, _ = model.smoother()
+        mu_sequence = mu_zo_smooth[:input_seq_len]
+
+    return np.nansum(mses) / np.sum(~np.isnan(mses))
+
+
+class SineSignalTest(unittest.TestCase):
+
+    def setUp(self):
+        self.threshold = 0.4
+
+    def test_lstm_CPU(self):
+        input_seq_len = 5
+        model = Sequential(
+            LSTM(1, 8, input_seq_len),
+            LSTM(8, 8, input_seq_len),
+            Linear(8 * input_seq_len, 1),
+        )
+        mse = lstm_test_runner(model, input_seq_len=input_seq_len)
+        self.assertLess(mse, self.threshold, "Error rate is higher than threshold")
+
+    def test_smoother_CPU(self):
+        input_seq_len = 24
+        num_features = 3
+        model = Sequential(
+            SLSTM(num_features + input_seq_len - 1, 32, 1),
```

Comment:
8 units are used in `cpp`

Reply:
I changed all py units to 8 for consistency